api/envoy/api/v2/core/base.proto

```diff
@@ -157,9 +157,31 @@ message HeaderValueOption {
   // Header name/value pair that this option applies to.
   HeaderValue header = 1 [(validate.rules).message.required = true];
 
-  // Should the value be appended? If true (default), the value is appended to
-  // existing values.
-  google.protobuf.BoolValue append = 2;
+  // Action to take when evaluating duplicate header keys.
+  enum HeaderAction {
+    option (gogoproto.goproto_enum_prefix) = false;
+    // Append new value of header to existing value.
+    APPEND = 0;
+    // Replace existing header value with new value.
+    REPLACE = 1;
+    // Ignore new header value and preserve existing value.
+    PRESERVE = 2;
+  }
+  oneof duplicate_action {
+    // Should the value be appended? If true (default), the value is appended to
+    // existing values.
+    //
+    // .. attention::
+    //
+    //   **This field is deprecated**. Set the
+    //   :ref:`action<envoy_api_field_core.HeaderValueOption.action>` field instead.
+    google.protobuf.BoolValue append = 2 [deprecated = true];
+
+    // Action to take when duplicate header values are encountered. By default,
+    // new values will be appended to the existing ones.
+    HeaderAction action = 3
+        [(validate.rules).message.required = true, (validate.rules).enum.defined_only = true];
```

Comment:
Provided we go forward with this implementation, does `message.required = true` make the field required?

I'm not sure because I think it's meant to apply to message types and HeaderAction is an enum. If it does, I think it's a problem because using a new Envoy with an older config (that doesn't know about this new field) will fail validation.

Reply:
I can make it optional. I was under the impression the oneof would allow this to be optional until we removed the deprecated field.

Reply:
I think with the oneof it won't actually be required, but I'm not sure. You could test it.